template for AKS with AAD integration based on 101-aks-advanced-network

t1agob opened the pull request:
### Best Practice Checklist
Check these items before submitting a PR... See the Contribution Guide for the full detail: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/README.md 

1. uri's compatible with all clouds (Stack, China, Government)
1. Staged artifacts use _artifactsLocation & _artifactsLocationSasToken
1. Use a parameter for resource locations with the defaultValue set to resourceGroup().location
1. Folder names for artifacts (nestedtemplates, scripts, DSC)
1. Use literal values for apiVersion (no variables)
1. Parameter files (GEN-UNIQUE for value generation and no "changemeplease" values)
1. $schema and other uris use https
1. Use uniqueString() whenever possible to generate names for resources.  While this is not required, it's one of the most common failure points in a deployment. 
1. Update the metadata.json with the current date

For details: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/best-practices.md

- [x] - Please check this box once you've submitted the PR if you've read through the Contribution Guide and best practices checklist.

### Changelog

* Added Azure AD support to an existing template in this repo (101-aks-advanced-network)

bmoore-msft commented:
This fails to deploy for me with the following:

5:29:41 PM - Error: Code=InvalidTemplateDeployment; Message=The template deployment failed with error: 'Authorization failed for template resource 'virtualNetwork/subnet-1/Microsoft.Authorization/ebeb5471-c434-5f28-a365-321bd47aaadf' of type 'Microsoft.Network/virtualNetworks/subnets/providers/roleA
ssignments'. The client '29d0108f-1d92-49b7-8f7a-b2d5c3f57f5c' with object id '29d0108f-1d92-49b7-8f7a-b2d5c3f57f5c' does not have permission to perform action 'Microsoft.Authorization/roleAssignments/write' at scope '/subscriptions/f42b4319-067a-4548-a650-33a1553b3a42/resourceGroups/template-bot-ge
n-artifacts/providers/Microsoft.Network/virtualNetworks/virtualNetwork/subnets/subnet-1/providers/Microsoft.Authorization/roleAssignments/ebeb5471-c434-5f28-a365-321bd47aaadf'.'.
The deployment validation failed

What perms does the existing SPN need on the sub (this one has contributor perms in the sub)

t1agob commented:
Hello Brian,
I have been able to deploy this template both in the resource group where the VNET is and in a different Resource Group.

I see in your error message that your AppId and ObjectId for Service Principal are the same and that is not possible I believe, right? These are parameters in the ARM Template should it should be pretty easy to fix.

_**"The client '29d0108f-1d92-49b7-8f7a-b2d5c3f57f5c' with object id '29d0108f-1d92-49b7-8f7a-b2d5c3f57f5c' does not have permission to perform action..."**_

Could this be the reason why you are having this error while deploying?

bmoore-msft commented:
That's the id for the service principal doing deployment, it's not a parameter to the deployment.

t1agob commented:
I branched out this template from an existing one in the repo (101-aks-advanced-network). I haven't changed anything related to network or permissions. Just added support for the Azure AD piece and added the necessary parameters. It should work without any special permissions. 

I'll try it in a different subscription with a user that is only a contributor and let you know if it worked.

t1agob commented:
Issue was identified as discussed. There is a step where we are assigning permissions to the Service Principal on the VNET. This is only possible if you are an **owner**. 

Added this information on the readme file so the users are aware. Maybe this should be fixed on the original template as well. I can do it if you want.

bmoore-msft commented on 101-aks-advanced-networking-aad/azuredeploy.json:
```diff
@@ -0,0 +1,267 @@
+            "name": "[parameters('resourceName')]",
+            "apiVersion": "2018-03-31",
+            "location": "[parameters('location')]",
+            "tags": {},
```
remove empty properties

bmoore-msft commented on 101-aks-advanced-networking-aad/azuredeploy.json:
```diff
@@ -0,0 +1,267 @@
+    "parameters": {
+        "resourceName": {
+            "type": "string",
+            "metadata": {
```
add a defaultValue suitable for all users

bmoore-msft commented on 101-aks-advanced-networking-aad/azuredeploy.json:
```diff
@@ -0,0 +1,267 @@
+        },
+        "dnsPrefix": {
+            "type": "string",
+            "metadata": {
```
add a defaultValue

bmoore-msft commented on 101-aks-advanced-networking-aad/azuredeploy.json:
```diff
@@ -0,0 +1,267 @@
+            "type": "Microsoft.Resources/deployments",
+            "name": "ClusterSubnetRoleAssignmentDeployment",
+            "apiVersion": "2017-05-10",
+            "subscriptionId": "[subscription().subscriptionId]",
```
remove this

bmoore-msft commented on 101-aks-advanced-networking-aad/azuredeploy.json:
```diff
@@ -0,0 +1,267 @@
+                        {
+                            "type": "Microsoft.Network/virtualNetworks/subnets/providers/roleAssignments",
+                            "apiVersion": "2017-05-01",
+                            "name": "[concat(parameters('existingVirtualNetworkName'), '/', parameters('existingSubnetName'), '/Microsoft.Authorization/', guid(resourceGroup().id, deployment().name))]",
```
use a parameter for the role name and add a defaultValue using the newGuid() function.

bmoore-msft commented on 101-aks-advanced-networking-aad/azuredeploy.parameters.json:
```diff
@@ -0,0 +1,42 @@
+            "value": "GET-PREREQ-subnetName"
+        },
+        "AAD_ClientAppID":{
+            "value": "GET-CLIENT-APP-ID-FROM-AZURE-AD"
```
the GETs here are not valid for CI/CD so just remove them - since we're skipping CI it won't matter
